### PR TITLE
SSO: Fix SSO Web APIs as per widl specifications

### PR DESCRIPTION
**Pull request by izaman**

### CODE REVIEW

**Comment by izaman** [comment]
@halton @ds-hwang can u plz review the above pull requests? major changes are done to match widl specifications, and also fix some critical issues..

File: examples/sso-advanced.html [hunk 1]
```diff
@@ -141,6 +141,69 @@
...
+  return null;
+}
+
+function _getAppSessionByGuiId(guiId) {
+  if (guiId < 0) return null;
+  for (var i = 0; i < g_app_identities.length; i++) {
+    var appIdent = g_app_identities[i];
+    if (appIdent != null) {
+      for (var j = 0; j < appIdent.sessions.length; j++) {
+        if (appIdent.sessions[i].guiId == guiId)
```
Comment (deqing) [inline examples/sso-advanced.html]: Should this be `j` instead of `i`?

File: examples/sso-advanced.html [hunk 2]
```diff
@@ -141,6 +141,69 @@
...
+  return null;
+}
+
+function _getAppSession(sessionobj) {
+  if (guiId < 0) return null;
+  for (var i = 0; i < g_app_identities.length; i++) {
+    var appIdent = g_app_identities[i];
+    if (appIdent != null) {
+      for (var j = 0; j < appIdent.sessions.length; j++) {
+        if (Object.is(appIdent.sessions[i].sessionObject, sessionobj))
```
Comment (deqing) [inline examples/sso-advanced.html]: ditto

File: examples/sso-advanced.html [hunk 3]
```diff
@@ -248,13 +319,17 @@
 function onCreateIdentity() {
   var info = getInfo();
   var res = tizen.sso.authService.createIdentity(info);
-  if (res.syncOpErrorMsg != null) {
-    _logData('Identity creation FAILED with error: ' + res.asyncOpErrorMsg);
-    return;
+  if (res != null) {
+    res.then(onCreateIdentityComplete, function(err) {
+      _logData('CreateIdentity failed: ' + err);});
```
Comment (deqing) [inline examples/sso-advanced.html]: Pls add one more indent for incomplete line

File: examples/sso-advanced.html [hunk 4]
```diff
@@ -431,45 +524,46 @@
...
+  if (appIdent == null) {
     _logData('identity NOT selected/found. Please select/create an identity first');
     return;
   }
-  _logData('identity found with JSId: ' + identobj.jsid);
-  var sessionobj = identobj.getSessionByJSId(document.getElementById('sess_options').value);
-  if (sessionobj == null) {
+  _logData('identity found with GuiId: ' + appIdent.guiId);
+  var appSession = _getAppSessionByGuiId(
+    document.getElementById('sess_options').value);
```
Comment (deqing) [inline examples/sso-advanced.html]: add indent for incomplete line

File: src/sso/sso.gyp [hunk 5]
```diff
@@ -9,6 +9,7 @@
       'variables': {
         'packages': [
           'libgsignon-glib',
+          'json-glib-1.0',
```
Comment (deqing) [inline src/sso/sso.gyp]: Order please

**Comment by izaman** [comment]
@deqing thanks for your comments.. i pushed the changes as per comments..

**Comment by izaman** [comment]
@halton @deqing can u plz merge the request now? :)

File: examples/sso-advanced.html [hunk 6]
```diff
@@ -141,6 +141,69 @@
...
+  for (var i = 0; i < g_app_identities.length; i++) {
+    var appIdent = g_app_identities[i];
+    if (appIdent.guiId == guiId) {
+      return appIdent;
+    }
+  }
+  return null;
+}
+
+function _getAppSession(sessionobj) {
```
Comment (halton) [inline examples/sso-advanced.html]: sessionObj

File: examples/sso-advanced.html [hunk 7]
```diff
@@ -301,73 +385,78 @@
...
 }
 
 function onSessionStateChanged(event) {
   var sessionobj = event.session;
-  _logData('Session with jsid ' + sessionobj.jsid + ' state has changed to '
-    + sessionobj.sessionState);
+  var appSession = _getAppSession(sessionobj);
+  if (appSession != null) {
+    _logData('Session state has changed to ' + sessionobj.sessionState +
+        ' for session with guiId ' + appSession.guiId);
```
Comment (halton) [inline examples/sso-advanced.html]: nit, align with string
```
_logData('Session state has changed to ' + sessionobj.sessionState +
         ' for session with guiId ' + appSession.guiId);
```

File: examples/sso-advanced.html [hunk 8]
```diff
@@ -378,31 +467,35 @@
...
-  if (info != null) identobj.updateInfo(info);
-
-  identobj.store().then(function(msg) {
-    _logData('store succeeded with resp: ' + JSON.stringify(msg));},
-    function(err) {_logData('store failed: ' + err);});
+  _logData('identity to store info: ' + appIdent.guiId);
+  if (info != null) {
+    appIdent.identityObject.storeWithInfo(info).then(function(msg) {
+        _logData('storeWithInfo succeeded with resp: ' + JSON.stringify(msg));},
+        function(err) {_logData('storeWithInfo failed: ' + err);});
```
Comment (halton) [inline examples/sso-advanced.html]: 2-space indent

File: examples/sso-advanced.html [hunk 9]
```diff
@@ -378,31 +467,35 @@
...
+  } else {
+    appIdent.identityObject.store().then(function(msg) {
+        _logData('store succeeded with resp: ' + JSON.stringify(msg));},
+        function(err) {_logData('store failed: ' + err);});
```
Comment (halton) [inline examples/sso-advanced.html]: 2-space indent

File: examples/sso-advanced.html [hunk 10]
```diff
@@ -431,45 +524,46 @@
...
+  if (appIdent == null) {
     _logData('identity NOT selected/found. Please select/create an identity first');
     return;
   }
-  _logData('identity found with JSId: ' + identobj.jsid);
-  var sessionobj = identobj.getSessionByJSId(document.getElementById('sess_options').value);
-  if (sessionobj == null) {
+  _logData('identity found with GuiId: ' + appIdent.guiId);
+  var appSession = _getAppSessionByGuiId(
+      document.getElementById('sess_options').value);
```
Comment (halton) [inline examples/sso-advanced.html]: squash to one line, we use 100 line limit with gslint, you could check `src/tools/check-coding-style`

File: examples/sso-advanced.html (hunk 4 above)
Comment (halton) [inline examples/sso-advanced.html]: instead, sqaush

File: examples/sso-advanced.html [hunk 11]
```diff
@@ -477,54 +571,54 @@
...
+  if (appIdent == null) {
     _logData('identity NOT selected/found. Please select/create an identity first');
     return;
   }
-  _logData('identity found with JSId: ' + identobj.jsid);
-  var sessionobj = identobj.getSessionByJSId(document.getElementById('sess_options').value);
-  if (sessionobj == null) {
+  _logData('identity found with GuiId: ' + appIdent.guiId);
+  var appSession = _getAppSessionByGuiId(
+    document.getElementById('sess_options').value);
```
Comment (halton) [inline examples/sso-advanced.html]: ditto

File: src/sso/sso_api.js [hunk 12]
```diff
@@ -41,33 +48,139 @@ function _isCommandInProgress(command) {
...
-    enumerable: true,
-    writable: true,
-    value: propObject[propertyKey]});
+function _addHiddenProperty(obj, propertyKey, propertyValue) {
+  _addProperty(obj, propertyKey, propertyValue, PropertyFlag.WRITEABLE);
+}
+
+function _addReadWriteProperty(obj, propertyKey, propertyValue) {
+  _addProperty(obj, propertyKey, propertyValue, PropertyFlag.ENUMERABLE |
+      PropertyFlag.WRITEABLE);
```
Comment (halton) [inline src/sso/sso_api.js]: can be squash

File: src/sso/sso_api.js [hunk 13]
```diff
@@ -41,33 +48,139 @@ function _isCommandInProgress(command) {
...
+
+function EventTargetInterface(eventListeners, isValidType) {
+  _addHiddenProperty(this, 'listeners', eventListeners);
+  _addHiddenProperty(this, 'isValidType', isValidType);
+}
+
+EventTargetInterface.prototype.addEventListener = function(type, callback) {
+  if (callback != null && typeof callback === 'function' &&
+      this.isValidType == type) {
+    if (this.listeners[type].indexOf(callback) != -1)
```
Comment (halton) [inline src/sso/sso_api.js]:
```
if (callback != null && typeof callback === 'function' &&
    this.isValidType == type && this.listeners[type].indexOf(callback) != -1)
```

File: src/sso/sso_api.js [hunk 14]
```diff
@@ -41,33 +48,139 @@ function _isCommandInProgress(command) {
...
+  if (callback != null && typeof callback === 'function' &&
+      this.isValidType == type) {
+    var index = this.listeners[type].indexOf(callback);
+    if (index >= 0)
+      this.listeners[type].slice(index, 1);
+  }
+};
+
+EventTargetInterface.prototype.dispatchEvent = function(event) {
+  var handled = true;
```
Comment (halton) [inline src/sso/sso_api.js]: declare it as close as first use

File: src/sso/sso_api.js [hunk 15]
```diff
@@ -41,33 +48,139 @@ function _isCommandInProgress(command) {
...
+};
+
+function _updateIdentityProps(src, dest) {
+  for (var prop in src) {
+    if (src.hasOwnProperty(prop) && src[prop] && dest.hasOwnProperty(prop))
+      dest[prop] = src[prop];
+  }
+}
+
+function _getSessionObj(identity, sessionJSId) {
```
Comment (halton) [inline src/sso/sso_api.js]: Wrong name, `sessionJsId` or `sessionId` or `sessionJavascriptId`

File: src/sso/sso_api.js [hunk 16]
```diff
@@ -41,33 +48,139 @@ function _isCommandInProgress(command) {
...
+    if (identity.sessions[j].method == method) {
+      session = identity.sessions[j];
+      break;
+    }
+  }
+  return session;
+}
+
+function _getSessionObjByJSId(sessionJSId) {
+  var session = null;
```
Comment (halton) [inline src/sso/sso_api.js]: declare as close as first use

File: src/sso/sso_api.js [hunk 17]
```diff
@@ -41,33 +48,139 @@ function _isCommandInProgress(command) {
...
+    if (identType == 'APPLICATION')
+      return IdentityType.APPLICATION;
+    else if (identType == 'NETWORK')
+      return IdentityType.NETWORK;
+  } else if (identType == IdentityType.APPLICATION) {
+    return IdentityType.APPLICATION;
+  } else if (identType == IdentityType.NETWORK) {
+    return IdentityType.NETWORK;
+  }
+  return IdentityType.WEB;
```
Comment (halton) [inline src/sso/sso_api.js]:
```
if (isNaN(identType)) {
  if (identType == 'APPLICATION')
    return IdentityType.APPLICATION;
  if (identType == 'NETWORK')
    return IdentityType.NETWORK;
}

return (identType == IdentityType.APPLICATION || identType == IdentityType.NETWORK) ?
       identType : IdentityType.WEB;
```

File: src/sso/sso_api.js [hunk 18]
```diff
@@ -102,91 +215,58 @@ var UserPromptPolicy = {
...
-      if (sess.jsid == sessionJSId) {
-        sessobj = sess;
-        break;
-      }
-    }
-  }
-  return sessobj;
+function IdentityInfo(info) {
+  var type = _convertToIdentType((info && info.hasOwnProperty('type') &&
+      info.type) || 'WEB');
```
Comment (halton) [inline src/sso/sso_api.js]: sqaush

File: src/sso/sso_api.js [hunk 19]
```diff
@@ -102,91 +215,58 @@ var UserPromptPolicy = {
...
+  _addReadWriteProperty(this, 'storeSecret',
+      Boolean((info && info.hasOwnProperty('storeSecret') && info.storeSecret) ||
+      false));
+  _addReadWriteProperty(this, 'realms', (info && info.hasOwnProperty('realms') &&
+      info.realms) || null);
+  _addReadWriteProperty(this, 'owner', (info && info.hasOwnProperty('owner') &&
+      info.owner) || null);
+  _addReadWriteProperty(this, 'accessControlList',
+      (info && info.hasOwnProperty('accessControlList') &&
+      info.accessControlList) || null);
```
Comment (halton) [inline src/sso/sso_api.js]: consider squash

File: src/sso/sso_api.js [hunk 20]
```diff
@@ -264,34 +349,53 @@ function handleAsyncCallSuccess(msg) {
...
+      identities.splice(identities.indexOf(identity), 1);
   }
   g_async_calls[msg.asyncOpId].reject(Error(msg.asyncOpErrorMsg));
   delete g_async_calls[msg.asyncOpId];
 }
 
+function handleCreateIdentity(msg) {
+  var identities = _sso.authService.identities;
+  var identity = _getIdentityObj(identities, 'jsid',
+      msg.responseData.identityJSId);
```
Comment (halton) [inline src/sso/sso_api.js]: squash

File: src/sso/sso_api.js [hunk 21]
```diff
@@ -264,34 +349,53 @@ function handleAsyncCallSuccess(msg) {
...
 function handleGetIdentity(msg) {
-  var identobj = _getIdentityObjByJSId(msg.responseData.identityJSId);
-  if (identobj) {
-    identobj.info = new IdentityInfo(msg.responseData.info,
-        msg.responseData.info.id, _convertToIdentType(msg.responseData.info.type),
-        msg.responseData.info.owner);
-    g_async_calls[msg.asyncOpId].resolve(identobj);
+  var identities = _sso.authService.identities;
+  var identity = _getIdentityObj(identities, 'jsid',
+      msg.responseData.identityJSId);
```
Comment (halton) [inline src/sso/sso_api.js]: ditto

File: src/sso/sso_api.js [hunk 22]
```diff
@@ -264,34 +349,53 @@ function handleAsyncCallSuccess(msg) {
...
 
+function handleCreateIdentity(msg) {
+  var identities = _sso.authService.identities;
+  var identity = _getIdentityObj(identities, 'jsid',
+      msg.responseData.identityJSId);
+  if (identity != null) {
+    g_async_calls[msg.asyncOpId].resolve(identity);
+  } else {
+    g_async_calls[msg.asyncOpId].reject(
+        Error('Identity object not found for create Identity'));
```
Comment (halton) [inline src/sso/sso_api.js]: ditto

File: src/sso/sso_api.js [hunk 23]
```diff
@@ -264,34 +349,53 @@ function handleAsyncCallSuccess(msg) {
...
+  var identity = _getIdentityObj(identities, 'jsid',
+      msg.responseData.identityJSId);
+  if (identity instanceof Identity) {
+    _updateIdentityProps(new IdentityInfo(msg.responseData.info), identity);
+    v8tools.forceSetProperty(identity, 'id', msg.responseData.info.id);
+    g_async_calls[msg.asyncOpId].resolve(identity);
   } else {
-    g_async_calls[msg.asyncOpId].reject(Error('Identity object not found for getidentity'));
+    g_async_calls[msg.asyncOpId].reject(
+        Error('Identity object not found for getidentity'));
```
Comment (halton) [inline src/sso/sso_api.js]: ditto

File: src/sso/sso_auth_session.cc [hunk 24]
```diff
@@ -226,8 +35,8 @@ void SSOAuthSession::ChallengeCb(GObject* source_object, GAsyncResult* res,
   SSOAsyncOp* async_op = reinterpret_cast<SSOAsyncOp*>(user_data);
   SSOAuthSession* session = reinterpret_cast<SSOAuthSession*>(
       async_op->user_data());
-
   GError* error = NULL;
+  gchar *data = NULL;
```
Comment (halton) [inline src/sso/sso_auth_session.cc]: `gchar* data`

File: src/sso/sso_auth_session.h [hunk 25]
```diff
@@ -66,6 +44,8 @@ class SSOAuthSession {
   virtual ~SSOAuthSession();
 
   int jsid() const { return jsid_; }
+  SignonAuthSession* session() const { return session_; }
+  std::string method() const { return method_; }
```
Comment (halton) [inline src/sso/sso_auth_session.h]: Look at where you use this getter, I'd suggest to define as `const std::string& `

File: src/sso/sso_identity.cc [hunk 26]
```diff
@@ -140,20 +140,33 @@ SSOIdentity::~SSOIdentity() {
   }
 }
 
-void SSOIdentity::HandleStartSession(const picojson::value& value) {
+void SSOIdentity::HandleGetSession(const picojson::value& value) {
   GError* error = NULL;
```
Comment (halton) [inline src/sso/sso_identity.cc]: declare error inside `if (!session)` condition

**Comment by halton** [comment]
Remove comma after [SSO]

If the spec changes can be summarized or has version info, please add it in git log.

**Comment by izaman** [comment]
@halton  pushed all changes as per your comments.

File: examples/sso-advanced.html (hunk 4 above)
Comment (halton) [inline examples/sso-advanced.html]: still not fixed

File: src/sso/sso_api.js [hunk 27]
```diff
@@ -1,14 +1,21 @@
-// Copyright (c) 2014 Intel Corporation. All rights reserved.
+// Copyright (c) 2014-2015 Intel Corporation. All rights reserved.
```
Comment (halton) [inline src/sso/sso_api.js]: nit, no need update existing copyright year for existing files.

File: src/sso/sso_auth_session.cc [hunk 28]
```diff
@@ -226,8 +35,8 @@ void SSOAuthSession::ChallengeCb(GObject* source_object, GAsyncResult* res,
   SSOAsyncOp* async_op = reinterpret_cast<SSOAsyncOp*>(user_data);
   SSOAuthSession* session = reinterpret_cast<SSOAuthSession*>(
       async_op->user_data());
-
   GError* error = NULL;
+  gchar* data = NULL;
```
Comment (halton) [inline src/sso/sso_auth_session.cc]: nit, declare it as close as first use

**Comment by halton** [comment]
lgtm after fixing the remaining three comments: 2 new and 1 old.

**Comment by izaman** [comment]
@halton done

**Comment by halton** [comment]
@izaman not yet, the old one is not fixed yet.

**Comment by izaman** [comment]
@halton can u plz point me in new changed version as I couldnt find it now

**Comment by halton** [comment]
In the conversation tab, there is unfolded one, @deqing and myself has comment  on that.

File: examples/sso-advanced.html (hunk 4 above)
Comment (izaman) [inline examples/sso-advanced.html]: That is squashed now.. plz check the latest commit

File: examples/sso-advanced.html [hunk 29]
```diff
@@ -431,45 +518,42 @@
...
+  var appIdent = _getAppIdentityByGuiId(document.getElementById('ident_options').value);
+  if (appIdent == null) {
     _logData('identity NOT selected/found. Please select/create an identity first');
     return;
   }
-  _logData('identity found with JSId: ' + identobj.jsid);
-  var sessionobj = identobj.getSessionByJSId(document.getElementById('sess_options').value);
-  if (sessionobj == null) {
+  _logData('identity found with GuiId: ' + appIdent.guiId);
+  var appSession = _getAppSessionByGuiId(document.getElementById('sess_options').value);
```
Comment (izaman) [inline examples/sso-advanced.html]: @halton plz check here

File: examples/sso-advanced.html (hunk 4 above)
Comment (halton) [inline examples/sso-advanced.html]: The code still look to me as
```
var appSession = _getAppSessionByGuiId(
  document.getElementById('sess_options').value);
```

What I meant is
```
var appSession = _getAppSessionByGuiId(document.getElementById('sess_options').value);
```

**Comment by izaman** [comment]
@halton hopefully now it is resolved :)

**Pull request merged**